test/e2e/testsuites/dynamically_provisioned_garbage_collection_replica_attachment_tester.go

```diff
@@ -0,0 +1,180 @@
+		Client: client,
+	}
+	csiNode, err := client.StorageV1().CSINodes().Get(context.Background(), nodes[0], metav1.GetOptions{})
+	framework.ExpectNoError(err, "Expected to successfully get the CSI node using the API.")
+
+	numReplicas, ok := driverNumReplicas(csiNode)
+	framework.ExpectEqual(ok, true, "Expected to find %s.", consts.DefaultDriverName)
+	framework.Logf("numReplicas is %d", numReplicas)
+
+	// Cordoning all but two nodes
+	for _, node := range nodes[2:] {
+		node := node
+		framework.Logf("Cordoning %s.", node)
+		testPod.SetNodeUnschedulable(node, true)
+		defer func() {
+			framework.Logf("Uncordoning %s", node)
+			testPod.SetNodeUnschedulable(node, false)
+		}()
+	}
+
+	tStorageClass, storageCleanup := t.StatefulSetPod.CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer storageCleanup()
+
+	labels := map[string]string{"group": "delete-for-replica-scaleup"}
+	tStatefulSet, cleanupStatefulSet := t.StatefulSetPod.SetupStatefulset(client, namespace, t.CSIDriver, schedulerName, numReplicas, &tStorageClass, labels)
+	isStatefulSetCleanup := false
+
+	// Wrap the stateful set cleanup; we want to call this at a specific time but also call it in case of failure.
+	wrappedCleanup := func() {
+		if !isStatefulSetCleanup {
+			cleanupStatefulSet(statefulSetCleanupTimeout)
+			isStatefulSetCleanup = true
+		}
+	}
+	defer wrappedCleanup()
+
+	tStatefulSet.Create()
+	// Check that AzVolumeAttachment resources were created correctly
+	ginkgo.By("Verifying that the stateful set attachments are successful.")
+	verifyStatefulSetAttachments := func() (bool, []v1beta1.AzVolumeAttachment, error) {
+		return t.allStatefulSetAttached(client, namespace, tStatefulSet)
+	}
+	failedAttachments, err := t.verifyAttachmentPolling(replicaCheckTimeout, verifyStatefulSetAttachments)
+	framework.ExpectEmpty(failedAttachments, "Expected there to be no failed attachments.")
+	framework.ExpectNoError(err, "Expected the stateful set to have successful attachments.")
+	framework.Logf("Stateful set replica attachments verified successfully.")
+
+	ginkgo.By("Creating a pod with volume.")
+	// Uncordon a node for the primary attachment
+	testPod.SetNodeUnschedulable(nodes[2], false)
+	framework.Logf("Uncordoning %s.", nodes[2])
+	tpod, cleanupFuncs := t.NewPod.SetupWithDynamicVolumes(client, namespace, t.CSIDriver, t.StorageClassParameters, schedulerName)
+	for _, cleanupFunc := range cleanupFuncs {
+		defer cleanupFunc()
+	}
+
+	ginkgo.By("Deploying the pod.")
+	tpod.Create()
+	defer tpod.Cleanup()
+	ginkgo.By("Checking that the pod is running.")
+	tpod.WaitForRunning()
+	t.NewPod.Name = tpod.Pod.Name
+
+	ginkgo.By("Verifying that new pod was not able to create any replica attachments.")
+	verifyPodAttachment := func() (bool, []v1beta1.AzVolumeAttachment, error) {
+		isVerified, _, failedAttachments, err := resources.VerifySuccessfulAzVolumeAttachments(t.NewPod, t.AzDiskClient, t.StorageClassParameters, client, namespace)
```

Comment:
Strange, `resources.VerifySuccessfulAzVolumeAttachments` doesn't seem to *ever* return an error. How does the test pass the `framework.ExpectError` on line 134 below? Oh, I see, the timeout error returned by `wait.PollImmediate` is what causes this check to pass.

Philosophically speaking, I don't believe test validation methods should return an error unless the error itself is meaningful to the test. Unexpected errors encountered during the test (like API server errors) should just raise the failure in the method validation itself. Expected error conditions should explicitly be validated likewise. Here, we are validating that a Pod was scheduled but no replicas created and so we should verify that the number of replicas is zero.

Reply:
I guess checking explicitly for a `context.DeadlineExceeded` error would also be sufficient here

Reply:
I was thinking about this on the way home. Rather than waiting for the absence of replicas (and a forced delay in the test), we could look for the event recorded on the pod indicating the lack of available nodes for replicas. See [`createReplicas` in `common.go#L2124`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/8cf65b5d12eb5e075b8e5ecc295c2df5576864b9/pkg/controller/common.go#L2124)

Reply:
Here's an example on StackOverflow of how to use `client-go` to get the events for a pod: https://stackoverflow.com/questions/69790386/how-can-i-get-events-messages-from-a-pod-from-kubernetes-using-client-go-api.

Reply:
I see what you're saying. You can see why polling for the non-existence of some property (for all time) is more difficult than looking for the existence of some property at a particular time.

I will investigate making this change. Thanks for the link and code reference.